[3단계] - 자동차 경주

Jo-GyuHyeon opened the pull request:
안녕하세요 리뷰어님 😄
3단계 구현 완료 되어 리뷰 요청 드립니다. 👶 

자동차 경주 기능 구현에 필요한 클래스들을 아래와 같이 정의 해보았습니다.
- [IO - UI 를 담당하는 패키지]
- [Car - 자동차 도메인]
- [Random - 전진 조건 제시 (랜덤)]
- [Track - 자동차 경주장]

감기조심하세요 🗡️ 
>  step2 리뷰 pr 반영이 안되어 compare 에 포함되는 것 같습니다. 😢 


---
추가 ➕ 
안녕하세요 리뷰어님 🙇 
리뷰를 부분적으로 적용해보았습니다. 😄 
(컨벤션, 가시성, 등)
랜덤 정수 생성하는 interface 는 단일 추상 메서드이기에 (Single Abstract Method) Functional interface 로 해보는 건 어떨까 싶어서 미리 적용 해보았습니다.

Jo-GyuHyeon committed:
refactor: 코드리뷰 반영

- companion object 컨벤션
- 테스트 SRP 원칙 위반 반영

Jo-GyuHyeon committed:
feat: UI 를 담당하는 로직 추가

- 입출력을 담당하는 클래스 작성
- 입출력에 대한 테스트 제외

Jo-GyuHyeon committed:
feat: 자동차 도메인 추가

- 자동차 도메인 테스트 추가

Jo-GyuHyeon committed:
feat: 자동차 이동 조건을 위한 무작위 값 생성 기능 추가

- 랜덤 범위 테스트 (최소, 최대값 경계 테스트)

Jo-GyuHyeon committed:
feat: 자동차 경주 실행 추가

malibinYun commented on src/main/kotlin/racingcar/Car.kt:
```diff
@@ -0,0 +1,13 @@
+package racingcar
+
+class Car {
+    var position: Int = 0
```
position을 생성자로부터 받아오는 것은 어떨까요?
테스트를 할 때 move()를 호출하지 않고도 원하는 거리로 배치된 자동차를 만들어 낼 수 있어 테스트할 때 용이해지겠네요 :)

malibinYun replied:
또, 외부에서 position 값을 변경시킬 수 있는 상태예요.
외부에서는 position에 직접 접근해서 값을 변경하지 못하게 막아보는 것은 어떨까요?

Jo-GyuHyeon replied:
position 을 생성자의 기본값을 지정할지 멤버 변수로써만 선언할지 고민을 했었는데 제한을 확실히 해야 된다고 생각했던것 같습니다.
position 의 접근제한 대해 깜빡해버렸네요. 😢

malibinYun commented on src/main/kotlin/racingcar/Car.kt:
```diff
@@ -0,0 +1,13 @@
+class Car {
+    var position: Int = 0
+
+    fun move(random: Int) {
+        if (random >= FOREWORD_CRITERIA) this.position++
+    }
```
랜덤한 값을 외부로부터 받아오게 만드셨군요! 👍👍

malibinYun commented on src/main/kotlin/racingcar/RandomStrategy.kt:
```diff
@@ -0,0 +1,13 @@
+package racingcar
+
+interface RandomStrategy {
```
랜덤 전략이 어떤 전략인지 좀 더 구체적으로 이름을 지어보는 것은 어떨까요?

malibinYun commented on src/main/kotlin/racingcar/RandomStrategy.kt:
```diff
@@ -0,0 +1,13 @@
+package racingcar
+
+interface RandomStrategy {
+    fun generateInt(): Int =
+        seed.random()
```
코틀린 컨벤션에서는 한 줄로 쓸 수 있다면, 한 줄로 쓰는 것을 권장합니다 :)
https://kotlinlang.org/docs/coding-conventions.html#functions

malibinYun replied:
또, interface에서 바로 구현하는 것 보다는,
interface는 숫자를 생성한다는 의미의 고수준의 추상화로 남겨두고,
랜덤에 대한 숫자를 반환하는 하위 구현체로 만들어 보는 것은 어떨까요?

Jo-GyuHyeon replied:
한줄 컨벤션을 배우게 되었습니다. 🙇 
interface 로 바로 구현하여 사용할 목적 보다는 default method 로 사용하려고 생각하였습니다.
interface 를 implement 하는 하위 구현체에서 overide에 대한 선택을 주고 싶었습니다. 😸 

지금의 경우에는 구현체 클래스를 선언하더라도 코드에 사용되지 않을것 같아 별도로 코드 작성은 하지 않았는데 (파라미터의 타입은 인터페이스 타입으로 정의 되었기 때문) 코드를 남겨 두는게 좋을까요❓

malibinYun replied:
자동차가 숫자를 받아서 움직이는 것이어서 `interface는 숫자를 생성한다는 의미의 고수준의 추상화`라는 의견을 드렸어요.
여기에 들어올 수 있는 숫자는 반드시 랜덤이 아닐 수 있는데요, 이 때에 고정 값을 넣는 전략이 생기는 경우, RandomStrategy 하위 구현체가 고정 값을 반환하는 것 자체가 
씌인 이름과 아귀가 맞지 않는다는 생각이 들었어요 :)

저는 interface는 구현 해야할 기능 목록만 가지고, 하위 구현체가 반드시 구현해야 한다고 생각해요. default method를 활용한다면 추상 클래스를 활용하는 편이에요 :)

Jo-GyuHyeon replied:
숫자를 생성하는 부분에서 무조건적으로 랜덤으로만 생성해야 해야 된다라는 갇힌 생각을 하게 된것 같습니다. 감사합니다.👍

malibinYun commented on src/test/kotlin/calculator/OperatorTest.kt:
```diff
@@ -2,16 +2,22 @@ package calculator
-import io.kotest.matchers.types.shouldBeInstanceOf
 
 internal class OperatorTest : StringSpec({
+    "올바른 사칙연산 기호" {
+        val symbols = Operator.values().map { it.symbol }
+        val expectSymbols = listOf("+", "-", "*", "/")
+
+        symbols shouldHaveSize 4
+        symbols shouldContainAll expectSymbols
+    }
+
```
사칙 기호들이 모두 있음을 테스트하는 것은 의미가 없어보여요.
특정 기호 문자열을 넣었을 때 원하는 연산자를 찾아내는지에 대한 테스트를 해보는 것은 어떨까요?

malibinYun reviewed (changes requested):
3단계 미션 고생 많으셨어요 :)
코틀린 확장함수를 활용해서 코드를 깔끔하게 잘 작성해주셨네요 👍
이전 미션 피드백 반영 사항은 다음 PR에 계속 반영해주시면 돼요 :)
compare에 계속 남아있어도 괜찮습니다. 오히려 제가 더 리뷰해드릴 수 있어서 좋아요!

생각해보면 좋을 점들에 코멘트 남겨두었어요.
피드백 반영하시구 다시 리뷰 요청 부탁드릴게요! :)

Jo-GyuHyeon committed:
refactor: PR 리뷰 반영

- single line function 컨벤션 적용
- Car Visibility modifiers (가시성) 리뷰 적용
- 2단계 계산기 테스트 케이스 리뷰 적용

malibinYun commented on src/main/kotlin/racingcar/Car.kt:
```diff
@@ -0,0 +1,14 @@
+package racingcar
+
+class Car(position: Int = 0) {
+    var position: Int = position
+        private set
```
초기 값을 세팅할 수도 있고 외부에서는 수정이 불가능해졌네요! 👍

malibinYun commented on src/main/kotlin/racingcar/RandomStrategy.kt:
```diff
@@ -0,0 +1,5 @@
+package racingcar
+
+fun interface RandomStrategy {
+    fun generateInt(): Int
+}
```
fun interface를 활용했으니 바로 람다를 활용할 수 있게 됐네요 👍

malibinYun commented on src/main/kotlin/racingcar/io/ResultView.kt:
```diff
@@ -0,0 +1,20 @@
+        RESULT("실행 결과");
+    }
+
+    fun printMessage(message: Message) =
+        println(message.context)
+
+    fun printResult(distance: Int) =
+        println(MARK.repeat(distance.inc()))
+
+    fun printNewLine() =
+        println()
```
이 부분들도 한 줄로 써도 좋겠네요 😀

malibinYun commented on src/test/kotlin/calculator/OperatorTest.kt:
```diff
@@ -3,14 +3,38 @@ package calculator
-    "사칙연산 기호가 아닌 경우 예외 발생" {
+    "덧셈 기호 (+) 의 연산자" {
         val plusSymbol = "+"
-        val wrongSymbol = "@"
+        val plusOperator = Operator.of(plusSymbol)
+
+        plusOperator shouldBe Operator.PLUS
+    }
+
+    "뺄셈 기호 (-) 의 연산자" {
+        val minusSymbol = "-"
+        val minusOperator = Operator.of(minusSymbol)
+
+        minusOperator shouldBe Operator.MINUS
+    }
+
+    "곱셈 기호 (*) 의 연산자" {
+        val multiSymbol = "*"
+        val multiOperator = Operator.of(multiSymbol)
 
-        Operator.of(plusSymbol).shouldBeInstanceOf<Operator>()
+        multiOperator shouldBe Operator.MULTI
+    }
+
+    "나누셈 기호 (/) 의 연산자" {
+        val divSymbol = "/"
+        val divOperator = Operator.of(divSymbol)
+
+        divOperator shouldBe Operator.DIV
+    }
+
```
Junit 에서 Parameterized Test라고 불리는 kotest의 data-driven test를 작성해보는 것은 어떨까요?

https://github.com/kotest/kotest/blob/master/doc/data_driven_testing.md

Jo-GyuHyeon replied:
오호! DDT (Data Driven Test) 를 알게되었습니다. 🙇‍♂️ 
Step 4 에 적용하여 PR 에 함께 반영해보겠습니다. ✊

malibinYun reviewed (approved):
3단계 미션 고생 많으셨어요!!
생각해보면 좋을 점들에 코멘트 남겨두었습니다 :)
피드백은 다음 미션에 반영 부탁드려요!!
미션 끝까지 화이팅입니다 ㅎ_ㅎ 💪
